plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Core/Instance/Flat.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Language.UntypedPlutusCore.Core.Instance.Flat where
+
+import           Language.UntypedPlutusCore.Core.Type
+
+import           Language.PlutusCore.Flat
+import           Language.PlutusCore.Universe
+
+import           Data.Word                            (Word8)
+import           Flat
+import           Flat.Decoder
+import           Flat.Encoder
+
+encodeTerm :: Word8 -> Encoding
+encodeTerm = eBits 4
+
+decodeTerm :: Get Word8
+decodeTerm = dBEBits8 4
+
+instance ( Closed uni
+         , uni `Everywhere` Flat
+         , Flat ann
+         , Flat name
+         ) => Flat (Term name uni ann) where
+    encode = \case
+        Var      ann n    -> encodeTerm 0 <> encode ann <> encode n
+        Delay    ann t    -> encodeTerm 1 <> encode ann <> encode t
+        LamAbs   ann n t  -> encodeTerm 2 <> encode ann <> encode n <> encode t
+        Apply    ann t t' -> encodeTerm 3 <> encode ann <> encode t <> encode t'
+        Constant ann c    -> encodeTerm 4 <> encode ann <> encode c
+        Force    ann t    -> encodeTerm 5 <> encode ann <> encode t
+        Error    ann      -> encodeTerm 6 <> encode ann
+        Builtin  ann bn   -> encodeTerm 7 <> encode ann <> encode bn
+
+    decode = go =<< decodeTerm
+        where go 0 = Var      <$> decode <*> decode
+              go 1 = Delay    <$> decode <*> decode
+              go 2 = LamAbs   <$> decode <*> decode <*> decode
+              go 3 = Apply    <$> decode <*> decode <*> decode
+              go 4 = Constant <$> decode <*> decode
+              go 5 = Force    <$> decode <*> decode
+              go 6 = Error    <$> decode
+              go 7 = Builtin  <$> decode <*> decode
+              go _ = fail "Failed to decode Term TyName Name ()"
+
+    size tm sz = 4 + sz + case tm of
+        Var      ann n    -> getSize ann + getSize n
+        Delay    ann t    -> getSize ann + getSize t
+        LamAbs   ann n t  -> getSize ann + getSize n + getSize t
+        Apply    ann t t' -> getSize ann + getSize t + getSize t'
+        Constant ann c    -> getSize ann + getSize c
+        Force    ann t    -> getSize ann + getSize t
+        Error    ann      -> getSize ann
+        Builtin  ann bn   -> getSize ann + getSize bn
+
+instance ( Closed uni
+         , uni `Everywhere` Flat
+         , Flat ann
+         , Flat name
+         ) => Flat (Program name uni ann) where
+    encode (Program ann v t) = encode ann <> encode v <> encode t
+    decode = Program <$> decode <*> decode <*> decode
```

Comment:
I was worried here because with CBOR we had to go to some trouble to make sure that unit annotations weren't serialised. However, it looks as if Flat does that automatically (although that has me wondering how it encodes a list of units).  It might be worth pointing that out somewhere.

Reply:
Huh, [so it does](http://hackage.haskell.org/package/flat-0.4.4/docs/src/Flat.Instances.Base.html#line-146). I guess Flat doesn't feel the need to be self-describing, so it can simply omit them...

Worth mentioning in a note somewhere, though.

Reply:
Ouch. I have no idea how it encodes a list of units. Do we need to handle that case?

Reply:
I have added a line in the note saying that flat encodes `()` as empty space.

Reply:
I don't think we need to worry about lists of annotations. In principle that could be a problem, but it isn't here.